File: Sources/PimTrayCore/Models/RoleScope.swift
import Foundation

public enum RoleScopeKind: String, Codable, Hashable, Sendable, CaseIterable {
    case entraDirectory, azureResource, group
}

public enum GroupAccess: String, Codable, Hashable, Sendable { case member, owner }

public enum RoleScope: Codable, Hashable, Sendable {
    case entraDirectory(roleDefinitionId: String, directoryScopeId: String)
    case azureResource(scope: String, roleDefinitionId: String)
    case group(groupId: String, accessId: GroupAccess)

    public var kind: RoleScopeKind {
        switch self {
        case .entraDirectory: .entraDirectory
        case .azureResource: .azureResource
        case .group: .group
        }
    }
}

public struct RoleKey: Codable, Hashable, Sendable {
    public let identityId: String
    public let tenantId: String
    public let scope: RoleScope

    public init(identityId: String, tenantId: String, scope: RoleScope) {
        self.identityId = identityId
        self.tenantId = tenantId
        self.scope = scope
    }

    public var tenantKey: TenantKey { TenantKey(identityId: identityId, tenantId: tenantId) }
}

File: windows/src/Elevate.Core/Models/RoleScope.cs
using System.Text.Json.Serialization;
using Elevate.Core;

namespace Elevate.Core.Models;

public enum RoleScopeKind { EntraDirectory, AzureResource, Group }

public enum GroupAccess { Member, Owner }

/// <summary>
/// What a role assignment applies to. Serialised exactly like Swift's synthesised enum coding:
/// a single-key object keyed by the case name, e.g.
/// <c>{"entraDirectory":{"roleDefinitionId":"…","directoryScopeId":"…"}}</c>.
/// </summary>
[JsonConverter(typeof(RoleScopeJsonConverter))]
public abstract record RoleScope
{
    [JsonIgnore]
    public abstract RoleScopeKind Kind { get; }
}

public sealed record EntraDirectoryScope(string RoleDefinitionId, string DirectoryScopeId) : RoleScope
{
    [JsonIgnore]
    public override RoleScopeKind Kind => RoleScopeKind.EntraDirectory;
}

public sealed record AzureResourceScope(string Scope, string RoleDefinitionId) : RoleScope
{
    [JsonIgnore]
    public override RoleScopeKind Kind => RoleScopeKind.AzureResource;
}

public sealed record GroupScope(string GroupId, GroupAccess AccessId) : RoleScope
{
    [JsonIgnore]
    public override RoleScopeKind Kind => RoleScopeKind.Group;
}

/// <summary>Identity + tenant + scope: the stable identity of one role a user can hold.</summary>
public sealed record RoleKey(string IdentityId, string TenantId, RoleScope Scope)
{
    [JsonIgnore]
    public TenantKey TenantKey => new(IdentityId, TenantId);
}
